query/src/main/java/org/infinispan/query/impl/LifecycleManager.java

```diff
@@ -93,8 +94,9 @@
    @Override
    public void cacheStarting(ComponentRegistry cr, Configuration cfg, String cacheName) {
       InternalCacheRegistry icr = cr.getGlobalComponentRegistry().getComponent(InternalCacheRegistry.class);
-      if (!icr.isInternalCache(cacheName)) {
-         cr.registerComponent(new ReflectionMatcher(null), ReflectionMatcher.class);
+      if (!icr.isInternalCache(cacheName) || icr.internalCacheHasFlag(cacheName, Flag.QUERYABLE)) {
```

Comment:
I think the ```internalCacheHasFlag(cacheName, Flag.QUERYABLE)``` check has to be done in the LifecycleManager of remote query too, for completeness, even if that is not used in this PR.